File: packages/db/prisma/seed.ts
import prisma from "../src/index";
import bcrypt from "bcryptjs";
import crypto from "crypto";

async function main() {
  console.log("Seeding...");

  // 1. Roles & Permissions
  console.log("Seeding Roles & Permissions...");

  const rolesData = [
    { name: "Admin", description: "Full system access" },
    { name: "Creator", description: "Can create and manage resources" },
    { name: "Viewer", description: "Can view shared resources" },
    { name: "Default", description: "Basic access" },
  ];

  for (const role of rolesData) {
    await prisma.role.upsert({
      where: { name: role.name },
      update: { description: role.description },
      create: role,
    });
  }

  // Assign basic permissions (Simplified)
  // In a real app, we'd have a mapping of Role -> Permissions.
  // For now, ensuring Roles exist is the priority.

  // 2. Create Admin User
  const adminEmail = "admin@dbms.local"; // Using .local to avoid confusion with real emails
  const adminPassword = "password123"; // Strong password in prod!

  const adminRole = await prisma.role.findUnique({ where: { name: "Admin" } });

  if (adminRole) {
    const hashedPassword = await bcrypt.hash(adminPassword, 10);
    const adminUser = await prisma.user.upsert({
      where: { email: adminEmail },
      update: {
        roleId: adminRole.id,
        username: "admin",
      },
      create: {
        email: adminEmail,
        username: "admin",
        name: "System Admin",
        password: hashedPassword,
        roleId: adminRole.id,
      },
    });
    console.log(adminUser);
    console.log(
      `Admin user seeded: username: admin / email: ${adminEmail} / password: ${adminPassword}`,
    );
  }

  // 3. Existing Data Cleanup (Optional - can be commented out if persistence is desired)
  // console.log("Cleaning up Dbs...");
  // await prisma.queryHistory.deleteMany({});
  // await prisma.savedQuery.deleteMany({});
  // await prisma.db.deleteMany({});

  // 4. Add Local Database if not exists
  const existingLocalDetails = await prisma.db.findFirst({
    where: { databaseName: "dbms_platform" },
  });

  // Replicate simple encryption from api utils to avoid circular deps
  function encrypt(text: string): string {
    const secret =
      process.env.JWT_SECRET || "fallback-secret-key-must-be-secure";
    const key = crypto.scryptSync(secret, "salt", 32);
    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv("aes-256-cbc", key, iv);
    let encrypted = cipher.update(text);
    encrypted = Buffer.concat([encrypted, cipher.final()]);
    return iv.toString("hex") + ":" + encrypted.toString("hex");
  }

  if (!existingLocalDetails) {
    await prisma.db.create({
      data: {
        databaseName: "dbms_platform",
        type: "postgres",
        config: {
          host: "localhost",
          port: 5432,
          user: "postgres",
          password: encrypt("postgres"),
          database: "dbms_platform",
        },
      },
    });
    console.log("Created Local Database connection");
  }

  console.log("Seeding completed.");
}

main()
  .then(async () => {
    await prisma.$disconnect();
  })
  .catch(async (e) => {
    console.error(e);
    await prisma.$disconnect();
    process.exit(1);
  });

File: backend/scripts/seed.py
"""
backend/scripts/seed.py

Database seed script - Creates initial roles, admin user, and local database connection.
Replaces the old Prisma seed.ts from packages/db.

Usage:
    cd backend
    python scripts/seed.py
"""

import sys
import os

# Add backend root to path so imports work
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from dotenv import load_dotenv
load_dotenv(os.path.join(os.path.dirname(__file__), '..', '.env'))

from models.metadata import SessionLocal, Role, User, Db, PrivilegeTypeModel, RolePrivilege
from services.privilege_service import privilege_service
from utils.crypto import encrypt
import uuid
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def seed():
    session = SessionLocal()

    try:
        # ── 1. Seed Roles ──────────────────────────────────────────────
        logger.info("Seeding Roles...")
        roles_data = [
            {"name": "Admin", "description": "Full system access"},
            {"name": "Creator", "description": "Can create and manage resources"},
            {"name": "Viewer", "description": "Can view shared resources"},
            {"name": "Default", "description": "Basic access"},
        ]

        for role_data in roles_data:
            existing = session.query(Role).filter(Role.name == role_data["name"]).first()
            if existing:
                existing.description = role_data["description"]
                logger.info(f"  Updated role: {role_data['name']}")
            else:
                role = Role(
                    id=str(uuid.uuid4()),
                    name=role_data["name"],
                    description=role_data["description"],
                )
                session.add(role)
                logger.info(f"  Created role: {role_data['name']}")

        session.commit()

        # ── 2. Seed Admin User ─────────────────────────────────────────
        logger.info("Seeding Admin User...")
        admin_email = "admin@dbms.local"
        admin_password = "password123"

        admin_role = session.query(Role).filter(Role.name == "Admin").first()
        if not admin_role:
            logger.error("Admin role not found! Cannot create admin user.")
            return

        existing_admin = session.query(User).filter(User.email == admin_email).first()
        if existing_admin:
            existing_admin.roleId = admin_role.id
            existing_admin.username = "admin"
            logger.info(f"  Updated admin user: {admin_email}")
        else:
            from passlib.context import CryptContext
            pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")
            hashed = pwd_context.hash(admin_password)

            admin_user = User(
                id=str(uuid.uuid4()),
                email=admin_email,
                username="admin",
                name="System Admin",
                password=hashed,
                roleId=admin_role.id,
            )
            session.add(admin_user)
            logger.info(f"  Created admin user: username=admin / email={admin_email} / password={admin_password}")

        session.commit()

        # ── 3. Seed Local Database Connection ──────────────────────────
        logger.info("Seeding Local Database Connection...")
        existing_db = session.query(Db).filter(Db.databaseName == "dbms_platform").first()

        if not existing_db:
            local_db = Db(
                id=str(uuid.uuid4()),
                databaseName="dbms_platform",
                type="postgres",
                config={
                    "host": "localhost",
                    "port": 5432,
                    "user": "postgres",
                    "password": encrypt("postgres"),
                    "database": "dbms_platform",
                },
            )
            session.add(local_db)
            session.commit()
            logger.info("  Created local database connection: dbms_platform")
        else:
            logger.info("  Local database connection already exists, skipping.")

        # ── 4. Seed Privilege Types ────────────────────────────────────
        logger.info("Seeding Privilege Types...")
        result = privilege_service.seed_defaults()
        logger.info(f"  Privilege types: {result['created']} created, {result['total']} total")

        # ── 5. Assign Default Privileges to Admin ──────────────────────
        logger.info("Assigning default privileges to Admin...")
        admin_role = session.query(Role).filter(Role.name == "Admin").first()
        if admin_role:
            default_privs = ["SQLLab_ACCESS", "CONNECTIONS_ACCESS"]
            for code in default_privs:
                priv = session.query(PrivilegeTypeModel).filter(PrivilegeTypeModel.code == code).first()
                if priv:
                    existing = session.query(RolePrivilege).filter(
                        RolePrivilege.roleId == admin_role.id,
                        RolePrivilege.privilegeTypeId == priv.id
                    ).first()
                    
                    if not existing:
                         privilege_service.assign_privilege({
                            "roleId": admin_role.id,
                            "privilegeTypeId": priv.id,
                            "resourceType": "SYSTEM"
                        })
                         logger.info(f"  Granted {code} to Admin")
        
        logger.info("Seeding completed successfully!")

    except Exception as e:
        session.rollback()
        logger.error(f"Seeding failed: {e}")
        raise
    finally:
        session.close()


if __name__ == "__main__":
    seed()
